### PR TITLE
fix(45501): VS Code removes spaces between empty object literals even if insertSpaceAfterOpeningAndBeforeClosingEmptyBraces is true

### LINKED ISSUE #45501: VS Code removes spaces between empty object literals even if `insertSpaceAfterOpeningAndBeforeClosingEmptyBraces` is true
Issue Type: <b>Bug</b>

I have enabled the `insertSpaceAfterOpeningAndBeforeClosingEmptyBraces` option in settings.json and I still see that empty object literals with a space between them (`{ }`) are collapsed to remove the space (`{}`).

As an extended example, doing the following works as expected:
```js
class Foo {
  doFoo() { }
}
```

But this does not:
```js
const foo = { }
```

Upon saving, the second item gets collapsed to `const foo = {}`

VS Code version: Code 1.59.0 (379476f0e13988d90fab105c5c19e7abc8b1dea8, 2021-08-04T23:13:12.822Z)
OS version: Windows_NT x64 10.0.19042
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz (8 x 2419)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|7.69GB (0.42GB free)|
|Process Argv|--crash-reporter-id 98bbd91f-17e6-483f-a746-35cacbab776f|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
better-comments|aar|2.1.0
bracket-pair-colorizer-2|Coe|0.2.1
vscode-eslint|dba|2.1.23
indent-rainbow|ode|7.5.0

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
pythonvspyt602:30300191
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyt639:30300192
pythontb:30283811
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
pythondataviewer:30285071
pythonvsuse255:30340121
vscod805cf:30301675
pythonvspyt200:30340761
vscextlangct:30333562
binariesv615:30325510
pythonvssor306:30344512
bridge0708:30335490
vstre464:30350172
bridge0723:30351747

```

</details>

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #45501

### CODE REVIEW

**Comment by andrewbranch** [comment]
Should object literals and blocks even be controlled by the same setting?

**Comment by a-tarasyuk** [comment]
@andrewbranch  I think it depends., the option name is `insertSpaceAfterOpeningAndBeforeClosingEmptyBraces` does that mean `{}` object literals? or only blocks? Maybe formatter should have two options `insertSpaceAfterOpeningAndBeforeClosingEmptyBlockBraces`, `insertSpaceAfterOpeningAndBeforeClosingEmptyObjectLiteralBraces`. I think it is a discussable question.

**Comment by andrewbranch** [comment]
What do other popular formatters do here?

**Comment by a-tarasyuk** [comment]
@andrewbranch I'm not sure that Prettier has the option to handle spaces between _empty_ braces`{}`. There is option `bracketSpacing`, however, it removes spaces between empty braces and leaves in non-empty object literals

```js
{
  "bracketSpacing": true
}
```
Before 
```ts
export function foo() { }
export const a = { };
export const b = { x: 1 };
```

After
```ts
export function foo() {}
export const a = {};
export const b = { x: 1 };
```

**Review by andrewbranch — APPROVED** [review]
I’m happy to go with this as more consistent, and we can split into two rules later if there’s demand for it.

File: src/services/formatting/rules.ts [hunk 1]
```diff
@@ -270,9 +269,9 @@ namespace ts.formatting {
...
             rule("SpaceBeforeCloseBrace", anyToken, SyntaxKind.CloseBraceToken, [isOptionEnabledOrUndefined("insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces"), isBraceWrappedContext], RuleAction.InsertSpace),
-            rule("NoSpaceBetweenEmptyBraceBrackets", SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken, [isNonJsxSameLineTokenContext, isObjectContext], RuleAction.DeleteSpace),
             rule("NoSpaceAfterOpenBrace", SyntaxKind.OpenBraceToken, anyToken, [isOptionDisabled("insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces"), isNonJsxSameLineTokenContext], RuleAction.DeleteSpace),
             rule("NoSpaceBeforeCloseBrace", anyToken, SyntaxKind.CloseBraceToken, [isOptionDisabled("insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces"), isNonJsxSameLineTokenContext], RuleAction.DeleteSpace),
+            rule("NoSpaceBetweenEmptyBraceBrackets", SyntaxKind.OpenBraceToken, SyntaxKind.CloseBraceToken, [isOptionDisabledOrUndefined("insertSpaceAfterOpeningAndBeforeClosingEmptyBraces"), isNonJsxSameLineTokenContext, isObjectContext], RuleAction.DeleteSpace),
```
Comment (andrewbranch) [inline src/services/formatting/rules.ts:274]: Is the default `undefined`? I feel like `const a = { }` is unconventional and shouldn’t be the default. That’s my immediate hesitation with controlling these with a single option.

Comment (a-tarasyuk) [reply src/services/formatting/rules.ts:274]: Yes, the default is `undefined`. Space between empty braces in object literals `{  }` will be removed.

Comment (andrewbranch) [reply src/services/formatting/rules.ts:274]: Ack, this whole **No**Space vs. **insertSpace** thing is so confusing... this looks good 👍

Comment (a-tarasyuk) [reply src/services/formatting/rules.ts:274]: Oke. Thanks. Wow, it was my 300th PR :).

**Pull request merged**